Dlsloan/nvme

dlsloan opened the pull request:
Miscellaneous changes for nvme testing using DEVS="/dev/nvme0n1 ... /dev/nvmeNn1" to run all tests.

dlsloan committed:
test_all: Remove loopback testing when DEVS are specified

loopback testing is not needed when we have real devices to run on, drop
loopback test if any DEVS are specified in test_all

dlsloan committed:
test2: Added test file generation

Remove hard-coded paths and replace with randomly generated local
files on first execution.

dlsloan committed:
test2: Replaced drop_caches command with echo 3 > sysfs

drop_commands is not a standard linux command, replace with the correct
sysfs write "echo 3 > /proc/sys/vm/drop_caches"

dlsloan committed:
md.py: Store devs and unused_devs in md class

Store devs and unsused_devs in md class to allow running grow tests on
specific devices.

dlsloan committed:
md.py: Changed journal to boolean type

Remove hard-coded ram drive for journaling, instead use the same
mechanisim "get_next_disk" as grow does to allow journal tests in any
environment.

dlsloan committed:
test3: Removed hard-coded ram-disk

Replace hard-coded ram disks with MDInstance's disk managment, getting
verification disk from md.get_next_disk, and iterating through md.devs
instead of all hard-coded ram-disks.

lsgunth commented on test3:
```diff
@@ -12,23 +12,24 @@ import pathlib
+        for dev in self._md.devs:
             sectors = int((sys_blk / dev / "size").read_text())
             count = (sectors << 9) >> 20
             subprocess.check_call(["dd", "if=/dev/zero", f"of=/dev/{dev}",
```
This is going to take a very long time to zero an nvme device... might be better to zero a max of 10MB now.... Looks like we're already only comparing the last 10M in the cmp command below already.

dlsloan replied:
Reviewing the code, it looks like my tests didn't actually run this (there's a bug in addition to how long it should have run for). I've fixed the of=/dev/{dev} -> of={dev} and added a new commit to address the sizing issue including superblock offsets, but is test_all not supposed to run some verification tests too?

lsgunth replied:
It's only run with the --verify option, which doesn't always pass so it's not in any of the test_all stuff.

iomartin commented on md.py:
```diff
@@ -183,6 +183,9 @@ def setup_from_args(self, args=None):
 
         args = parser.parse_args(args)
 
+        if isinstance(args.devs, str):
```
This looks fine, but to be more general this could go inside `_add_action()` (just check for `action.nargs == '+'`)

lsgunth replied:
I don't think the environment variable handles that correctly....

dlsloan replied:
Does this mean the way I have it is ok?

iomartin commented on md.py:
```diff
@@ -91,36 +91,41 @@ def _create_loop_disks(self, count, size):
         return ret
 
-    def setup(self, level=5, disks=None, ram_disks=None, loop_disks=None,
+    def setup(self, level=5, devs=None, ndisks=None, disk_type='ram',
```
How about making the default for ndisks be 0?

dlsloan replied:
Should probably be 3

dlsloan replied:
Leaving as it was in the original, feature changes can be a separate pr if they are needed.

iomartin commented on md.py:
```diff
@@ -154,10 +159,11 @@ def setup_from_args(self, args=None):
-        group.add_argument("--loop-disks", type=int,
-                           help="use loop devices instead of ram disks")
+        group.add_argument("--disk-type", default="default",
+                           choices=["default", "ram", "dev", "loopback"],
```
Do we really the "default" option? We can just check if disk_type is unset and then decide if we need to use ram or dev

iomartin commented on md.py:
```diff
@@ -184,20 +190,21 @@ def setup_from_args(self, args=None):
+            args.disk_type = 'ram' if args.devs is None else 'dev'
 
-        if not args.devs and not args.loop_disks and not args.disks:
+        if not args.disks:
```
If `args.disks == 0 and args.disk_type == 'dev'`, then I think it might be better to set `args.disks = len(args.devs)`

dlsloan replied:
There are lots of tests that use extra disks for things, we don't want to default to len(args.devs) ever or that will break things

iomartin replied:
Well, it just seems odd that if you pass a bunch of devs they aren't all used unless --disks is set as well

dlsloan replied:
when running batch tests you set the DEVS env once for all tests so it has to contain the maximum number of disks for any test. Some tests use one of the unused disks as a jornal dev, and test3 uses one as a verification disk. MDInstance has no way of knowing how these will be used, and since the default ndisks is used for both journaling and test3 we cannot use all of the devs provided by args.devs.

dlsloan replied:
Oh wait, on closer inspection args.disks is basically just a 0 check, so the default args is mute point since it is 3 which will not affect this.  Did you mean if someone specifies --disks 0 it should pull the len from devs or use the default 3 if none are present? I'm not completely opposed to that, but it does make it so ram/loop tests can run with journaling or test3 verification enabled and DEVS tests cannot.

iomartin replied:
> when running batch tests you set the DEVS env once for all tests so it has to contain the maximum number of disks for any test. Some tests use one of the unused disks as a jornal dev, and test3 uses one as a verification disk. MDInstance has no way of knowing how these will be used, and since the default ndisks is used for both journaling and test3 we cannot use all of the devs provided by args.devs.

If we need some devices to be used for special cases, wouldn't it be better to do so explicitly? Using DEVS this way may only make sense for batch testing, and even then it still feels odd to me. We should at least make the help clearer about this.

> Oh wait, on closer inspection args.disks is basically just a 0 check, so the default args is mute point since it is 3 which will not affect this. Did you mean if someone specifies --disks 0 it should pull the len from devs or use the default 3 if none are present? I'm not completely opposed to that, but it does make it so ram/loop tests can run with journaling or test3 verification enabled and DEVS tests cannot.

I hadn't really thought about this but this might not be too bad. I don't see a problem with being unable to run a few tests with DEVS this way as long as we capture that error and error out nicely.

dlsloan replied:
How about just renaming DEVS to AVAILABLE_DEVS? That way it's more clear they won't all be used, and doesn't change the way tests work when running with or without explicit devs.

lsgunth replied:
I would say AVAILABLE_DEVS is too long for an environment variable and not really necessary.

dlsloan replied:
Ok, I'm leaving all this as close to the original as possible (drop args.disks default).  I don't actually object to changing the way things work, but this pr is about getting all tests to work with a DEVS=... env arg. Changes to test behavious can be done in a separate pr.

iomartin commented on md.py:
```diff
@@ -154,10 +159,11 @@ def setup_from_args(self, args=None):
 
         group = parser.add_mutually_exclusive_group()
-        group.add_argument("-d", "--disks", "--ram-disks", type=int,
+        group.add_argument("-d", "--disks", type=int, default=3,
```
It might be better to no set the default explicitly here and just let it be handled below like you have.

dlsloan replied:
Most tests, leave this set to it's default value

iomartin commented on md.py:
```diff
@@ -228,16 +231,22 @@ def get_disks(self):
+                sectors = int((self._sysfs / "rd0" / "block" / "size").read_text())
+                self.devs.append(self._create_loop_disk(n, sectors << 9))
+            else:
+                raise MDInvalidArgumentError("Can't grow array further with out using loop or ram disks")
```
s/with out/without

iomartin commented on md.py:
```diff
@@ -186,7 +189,8 @@ def setup_from_args(self, args=None):
                             help="cache size")
-        parser.add_argument("--journal", help="journal type")
+        parser.add_argument("--journal", action='store_true',
+                            help="journal type")
```
The help message needs to be fixed now as well

dlsloan committed:
test3: Limit number of bytes zero'd in disks

When using phisical devices instead of ram-disks the size can be much
larger and speed lower, making zeroing each array disk in it's entirety
unpractical. This patch reduces the number of bytes zero'd to the
smallest number which is guarenteed to cover all data verified for
reasonably sized arrays.

iomartin reviewed (approved):
Looks good to me

lsgunth commented on md.py:
```diff
@@ -185,31 +185,35 @@ def setup(self, level=5, devs=None, ndisks=None, disk_type=None,
         if disk_type is None:
             disk_type = "dev" if devs else "ram"
 
+        self.na_devs = []
```
This na_devs seems to be unused....

dlsloan replied:
see line 275

lsgunth commented on md.py:
```diff
@@ -210,6 +211,9 @@ def setup(self, level=5, devs=None, ndisks=None, disk_type=None,
             raise MDInvalidArgumentError(f"Unknown disk_type: {disk_type}")
 
+        if journal:
+            journal = self.get_next_disk(True)
```
Why not just put this below?

dlsloan replied:
haha, that makes much more sense

lsgunth commented on md.py:
```diff
@@ -266,17 +270,28 @@ def get_disks(self):
+        if non_array:
+            dev_list = self.na_devs
+        else:
+            dev_list = self.devs
```
Oh, this is confusing.... I guess you're adding to it here.... But it never appears to be read?

Would be much more sensible to add to the list at the end of the function with an if statement instead of twice in the middle and returning the last....

dlsloan replied:
only the length is read currently for creating new ram disks, I could have just made it a counter in the non-array case, but martin and I have been talking a bit about changing the default DISKS arg to use all DEVS without breaking journal/verify/grow tests which will use it more.

lsgunth replied:
I don't really follow that. I'm just saying it should be something like this: 

```python3
def _get_next_disk(self):
    if len(self.extra_devs):
        new_dev = self.extra_devs.pop()
    else:
        ...
    return new_dev

def get_special_disk(self):
    new_dev = self.__get_next_disk()
    self.special_devs.append(new_dev)
    return new_dev

def grow(self):
    self.wait()
    dev = self._get_next_disk(self)
    self.devs.append(dev)
    ...
```

Besides that I think the way your doing it makes sense.

dlsloan replied:
ah, got you

lsgunth reviewed (approved):
Looks good to me, thanks
